numba/_lapack.c

```diff
@@ -0,0 +1,1035 @@
+        case 'd':
+            return (F_INT)(*((double *) val));
+        case 'c':
+            return (F_INT)(*((npy_complex64 *)val)).real;
+        case 'z':
+            return (F_INT)(*((npy_complex128 *)val)).real;
+        default:
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_SetString(PyExc_ValueError,\
+                            "invalid kind in cast");
+            PyGILState_Release(st);
+        }
+    }
+    return -1;
+}
+
+// real space eigen systems info from dgeev/sgeev
+NUMBA_EXPORT_FUNC(F_INT)
+numba_raw_rgeev(char kind, char jobvl, char jobvr,
+                F_INT n, void *a, F_INT lda, void *wr, void *wi, void *vl,
+                F_INT ldvl, void *vr, F_INT ldvr, void *work, F_INT lwork,
+                F_INT * info)
+{
+    void *raw_func = NULL;
+
+    switch (kind)
+    {
+        case 's':
+            raw_func = get_clapack_sgeev();
+            break;
+        case 'd':
+            raw_func = get_clapack_dgeev();
+            break;
+        default:
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_SetString(PyExc_ValueError,\
+                            "invalid kind of real space *geev call");
+            PyGILState_Release(st);
+        }
+        return -1;
+    }
+    if (raw_func == NULL)
+        return -1;
+
+    (*(rgeev_t) raw_func)(&jobvl, &jobvr, &n, a, &lda, wr, wi, vl, &ldvl, vr,
+                          &ldvr, work, &lwork, info);
+    return 0;
+}
+
+// real space eigen systems info from dgeev/sgeev
+// as numba_raw_rgeev but the allocation and error handling is done for the user
+// Args are as per LAPACK.
+NUMBA_EXPORT_FUNC(F_INT)
+numba_ez_rgeev(char kind, char jobvl, char jobvr,
+               F_INT n, void *a, F_INT lda, void *wr, void *wi, void *vl,
+               F_INT ldvl, void *vr, F_INT ldvr)
+{
+    F_INT info = 0;
+    size_t base_size = -1;
+    // find the function to call, decide on a base type size
+    switch (kind)
+    {
+        case 's':
+            base_size = sizeof(float);
+            break;
+        case 'd':
+            base_size = sizeof(double);
+            break;
+        default:
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_SetString(PyExc_ValueError,\
+                            "Invalid kind in numba_ez_rgeev");
+            PyGILState_Release(st);
+        }
+        return -1;
+    }
+    F_INT lwork = -1;
+    void * work = malloc(base_size);
+    numba_raw_rgeev(kind, jobvl, jobvr, n, a, lda, wr, wi, vl, ldvl,
+                    vr, ldvr, work, lwork, &info);
+    lwork = cast_from_X(kind, work);
+    free(work);
+    if(info < 0)
+    {
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_Format(PyExc_ValueError,\
+                         "LAPACK Error: on input %d\n", -info);
+            PyGILState_Release(st);
+        }
+        return -1;
+    }
+    work = malloc(base_size * lwork);
```

Comment:
Rather than `malloc` / `free`, I'd rather call `PyMem_RawMalloc` / `PyMem_RawFree`, at least on non-tiny allocations, as it makes the allocation visible to e.g. the tracemalloc module.